base/dates/types.jl

```diff
@@ -109,22 +109,22 @@ function DateTime(periods::Period...)
     y = Year(1); m = Month(1); d = Day(1)
     h = Hour(0); mi = Minute(0); s = Second(0); ms = Millisecond(0)
     for p in periods
-        typeof(p) <: Year && (y = p)
-        typeof(p) <: Month && (m = p)
-        typeof(p) <: Day && (d = p)
-        typeof(p) <: Hour && (h = p)
-        typeof(p) <: Minute && (mi = p)
-        typeof(p) <: Second && (s = p)
-        typeof(p) <: Millisecond && (ms = p)
+        typeof(p) <: Year && (y = p::Year)
+        typeof(p) <: Month && (m = p::Month)
+        typeof(p) <: Day && (d = p::Day)
+        typeof(p) <: Hour && (h = p::Hour)
+        typeof(p) <: Minute && (mi = p::Minute)
+        typeof(p) <: Second && (s = p::Second)
+        typeof(p) <: Millisecond && (ms = p::Millisecond)
```

Comment:
I thinks so

```
julia> f1(x::ANY) = typeof(x) <: Integer
f1 (generic function with 1 method)

julia> f2(x::ANY) = isa(x, Integer)
f2 (generic function with 1 method)

julia> @code_llvm f1(1)

define i1 @julia_f1_22308(%jl_value_t*) {
top:
  %1 = alloca [4 x %jl_value_t*], align 8
  %.sub = getelementptr inbounds [4 x %jl_value_t*], [4 x %jl_value_t*]* %1, i64 0, i64 0
  %2 = getelementptr [4 x %jl_value_t*], [4 x %jl_value_t*]* %1, i64 0, i64 2
  %3 = bitcast [4 x %jl_value_t*]* %1 to i64*
  store i64 4, i64* %3, align 8
  %4 = getelementptr [4 x %jl_value_t*], [4 x %jl_value_t*]* %1, i64 0, i64 1
  %5 = load i64, i64* bitcast (%jl_value_t*** @jl_pgcstack to i64*), align 8
  %6 = bitcast %jl_value_t** %4 to i64*
  store i64 %5, i64* %6, align 8
  store %jl_value_t** %.sub, %jl_value_t*** @jl_pgcstack, align 8
  store %jl_value_t* null, %jl_value_t** %2, align 8
  %7 = getelementptr [4 x %jl_value_t*], [4 x %jl_value_t*]* %1, i64 0, i64 3
  store %jl_value_t* null, %jl_value_t** %7, align 8
  %8 = getelementptr inbounds %jl_value_t, %jl_value_t* %0, i64 -1, i32 0
  %9 = bitcast %jl_value_t** %8 to i64*
  %10 = load i64, i64* %9, align 8
  %11 = and i64 %10, -16
  %12 = inttoptr i64 %11 to %jl_value_t*
  store %jl_value_t* %12, %jl_value_t** %2, align 8
  store %jl_value_t* inttoptr (i64 140225665290768 to %jl_value_t*), %jl_value_t** %7, align 8
  %13 = call %jl_value_t* @jl_f_subtype(%jl_value_t* null, %jl_value_t** %2, i32 2)
  %14 = bitcast %jl_value_t* %13 to i8*
  %15 = load i8, i8* %14, align 1
  %16 = and i8 %15, 1
  %17 = icmp ne i8 %16, 0
  %18 = load i64, i64* %6, align 8
  store i64 %18, i64* bitcast (%jl_value_t*** @jl_pgcstack to i64*), align 8
  ret i1 %17
}

julia> @code_llvm f2(1)

define i1 @julia_f2_22312(%jl_value_t*) {
top:
  %1 = alloca [4 x %jl_value_t*], align 8
  %.sub = getelementptr inbounds [4 x %jl_value_t*], [4 x %jl_value_t*]* %1, i64 0, i64 0
  %2 = getelementptr [4 x %jl_value_t*], [4 x %jl_value_t*]* %1, i64 0, i64 2
  %3 = bitcast [4 x %jl_value_t*]* %1 to i64*
  store i64 4, i64* %3, align 8
  %4 = getelementptr [4 x %jl_value_t*], [4 x %jl_value_t*]* %1, i64 0, i64 1
  %5 = load i64, i64* bitcast (%jl_value_t*** @jl_pgcstack to i64*), align 8
  %6 = bitcast %jl_value_t** %4 to i64*
  store i64 %5, i64* %6, align 8
  store %jl_value_t** %.sub, %jl_value_t*** @jl_pgcstack, align 8
  %7 = getelementptr [4 x %jl_value_t*], [4 x %jl_value_t*]* %1, i64 0, i64 3
  store %jl_value_t* %0, %jl_value_t** %2, align 8
  store %jl_value_t* inttoptr (i64 140225665290768 to %jl_value_t*), %jl_value_t** %7, align 8
  %8 = call %jl_value_t* @jl_f_isa(%jl_value_t* null, %jl_value_t** %2, i32 2)
  %9 = bitcast %jl_value_t* %8 to i8*
  %10 = load i8, i8* %9, align 1
  %11 = and i8 %10, 1
  %12 = icmp ne i8 %11, 0
  %13 = load i64, i64* %6, align 8
  store i64 %13, i64* bitcast (%jl_value_t*** @jl_pgcstack to i64*), align 8
  ret i1 %12
}
```
And I would be surprised if the first version is more efficient (although I would also be surprised if they differ by a lot)

I'm mainly asking because we have `isa` for this and it's shorter.